lib/api/objectDelete.js

```diff
@@ -24,48 +28,124 @@ export default function objectDelete(authInfo, request, log, cb) {
     }
     const bucketName = request.bucketName;
     const objectKey = request.objectKey;
+    let reqVersionId = request.query ? request.query.versionId : undefined;
+
+    if (reqVersionId && reqVersionId !== 'null') {
+        try {
+            reqVersionId = VID.decrypt(reqVersionId);
+        } catch (exception) { // eslint-disable-line
+            return cb(errors.InvalidArgument.customizeDescription(
+                    'Invalid version id specified'), null);
+        }
+    }
+
     const valParams = {
         authInfo,
         bucketName,
         objectKey,
+        versionId: reqVersionId || undefined,
         requestType: 'objectDelete',
         log,
     };
-    return services.metadataValidateAuthorization(valParams,
-        (err, bucket, objMD) => {
-            const corsHeaders = collectCorsHeaders(request.headers.origin,
-                request.method, bucket);
-            if (err) {
-                log.debug('error processing request', {
-                    error: err,
-                    method: 'metadataValidateAuthorization',
-                });
-                return cb(err, corsHeaders);
-            }
+
+    const canonicalID = authInfo.getCanonicalID();
+    let bucketMD = undefined;
+    let objectMD = undefined;
+    let corsHeaders = undefined;
+    let removeDeleteMarker = false;
+    let deleteOptions = undefined;
+    return async.waterfall([
+        callback => services.metadataValidateAuthorization(valParams, callback),
```

Comment:
I refactored objectDelete and may have taken care of this, let me know if you still see any issue with the cors headers.